File: packages/tui/src/form/builder/status.ts
import { progressFormStep, spinFormStep, streamFormStep, taskFormStep } from '#tui/form/builder/status/index';
import type { FormBuilder } from '#tui/form/builder/index';
import type { MaybePromise } from '#tui/types';
import type { Logger, Progress, TaskDefinition } from '#tui/status';

export type StatusBuilderMethods = {
	progress(this: FormBuilder, total: number, message?: string, name?: string): FormBuilder;
	progress<T, R>(this: FormBuilder, label: string, steps: Iterable<T> | number, callback: (step: T | number, bar: Progress) => MaybePromise<R>, hint?: string, name?: string): FormBuilder;
	spin<T>(this: FormBuilder, callback: () => MaybePromise<T>, message?: string, name?: string): FormBuilder;
	spin<T>(this: FormBuilder, message: string, callback: () => MaybePromise<T>, name?: string): FormBuilder;
	task<T>(this: FormBuilder, definition: TaskDefinition<T>, name?: string): FormBuilder;
	stream(this: FormBuilder, name?: string): FormBuilder;
	stream(this: FormBuilder): FormBuilder;
	stream(this: FormBuilder, source: AsyncIterable<string> | Iterable<string>, name?: string): FormBuilder;
	task<T>(this: FormBuilder, label: string, callback: (logger: Logger) => MaybePromise<T>, limit?: number, keepSummary?: boolean, subLabel?: string, name?: string): FormBuilder;
};

export const statusBuilderMethods: StatusBuilderMethods & ThisType<FormBuilder> = {
	progress: progressFormStep,
	spin: spinFormStep,
	stream: streamFormStep,
	task: taskFormStep,
};

File: provision/go/internal/form/status.go
package form

import "github.com/oullin/tui/internal/status"

func (builder *Builder) Progress(total int, message string, name ...string) *Builder {
	return builder.AddSideEffect(func() error {
		bar := status.NewProgress(total, message, "")
		bar.Start()
		bar.Finish()
		return nil
	}, name...)
}

func (builder *Builder) Spin(message string, callback func() (any, error), name ...string) *Builder {
	return builder.Add(func(*Responses, any, string) (any, error) {
		return status.Spin(message, callback)
	}, name...)
}

func (builder *Builder) Stream(name ...string) *Builder {
	return builder.Add(func(*Responses, any, string) (any, error) {
		return status.NewStream(), nil
	}, name...)
}

func (builder *Builder) Task(label string, callback func(*status.Logger) (any, error), limit int, keepSummary bool, subLabel string, name ...string) *Builder {
	return builder.Add(func(*Responses, any, string) (any, error) {
		return status.Task(label, callback, limit, keepSummary, subLabel)
	}, name...)
}
